neqo-transport/src/recovery/sent.rs

```diff
@@ -0,0 +1,382 @@
+#[derive(Debug, Default)]
+pub struct SentPackets {
+    /// The collection.
+    packets: BTreeMap<u64, SentPacket>,
+}
+
+impl SentPackets {
+    pub fn len(&self) -> usize {
+        self.packets.len()
+    }
+
+    pub fn track(&mut self, packet: SentPacket) {
+        self.packets.insert(packet.pn, packet);
+    }
+
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = &mut SentPacket> {
+        self.packets.values_mut()
+    }
+
+    /// Take values from a specified ranges of packet numbers.
+    /// The values returned will be reversed, so that the most recent packet appears first.
+    /// This is because ACK frames arrive with ranges starting from the largest acknowledged
+    /// and we want to match that.
+    pub fn take_ranges<R>(&mut self, acked_ranges: R) -> Vec<SentPacket>
+    where
+        R: IntoIterator<Item = RangeInclusive<PacketNumber>>,
+        R::IntoIter: ExactSizeIterator,
+    {
+        let mut result = Vec::new();
+        // Remove all packets. We will add them back as we don't need them.
+        let mut packets = std::mem::take(&mut self.packets);
+        for range in acked_ranges {
+            // For each acked range, split off the acknowledged part,
+            // then split off the part that hasn't been acknowledged.
+            // This order works better when processing ranges that
+            // have already been processed, which is common.
+            let mut acked = packets.split_off(range.start());
+            let keep = acked.split_off(&(*range.end() + 1));
+            self.packets.extend(keep);
+            result.extend(acked.into_values().rev());
+        }
+        self.packets.extend(packets);
+        result
+    }
+
+    /// Empty out the packets, but keep the offset.
+    pub fn drain_all(&mut self) -> impl Iterator<Item = SentPacket> {
+        std::mem::take(&mut self.packets).into_values()
+    }
+
+    /// See `LossRecoverySpace::remove_old_lost` for details on `now` and `cd`.
+    /// Returns the number of ack-eliciting packets removed.
+    pub fn remove_expired(&mut self, now: Instant, cd: Duration) -> usize {
+        let mut it = self.packets.iter();
+        // If the first item is not expired, do nothing (the most common case).
+        if it.next().map_or(false, |(_, p)| p.expired(now, cd)) {
+            // Find the index of the first unexpired packet.
+            let to_remove = if let Some(first_keep) =
+                it.find_map(|(i, p)| if p.expired(now, cd) { None } else { Some(*i) })
+            {
+                // Some packets haven't expired, so keep those.
+                let keep = self.packets.split_off(&first_keep);
+                std::mem::replace(&mut self.packets, keep)
+            } else {
+                // All packets are expired.
+                std::mem::take(&mut self.packets)
+            };
+            to_remove
+                .into_values()
+                .filter(SentPacket::ack_eliciting)
+                .count()
```

Comment:
The caller only needed a count, so I decided to return a count.  That seems to have improved benchmarks a noticeable amount too.